### ISSUE TITLE
Fix S2259 FN: Linq 'XxxOrDefault' extensions should create null and not-null constraints

### ISSUE THREAD

**Issue by pavel-mikula-sonarsource** [SonarSource/sonar-dotnet#3290]
### Description

As reported on [this community thread](https://community.sonarsource.com/t/c-missing-simple-possible-null-references/22784) Linq `FirstOrDefault()` extension methods does not trigger S2259 "'item' is null on at least one execution path."

All OrDefault Linq extensions should add null and not-null constraints to SE.

### Repro steps

```C#
string DoSomething(IEnumerable<object> list)
{
    var item = list.FirstOrDefault();
    return item.ToString(); // False Negative, item could be null here
}
```

### APIs to track:
* ElementAt()
* ElementAtOrDefault()
* First()
* FirstOrDefault()
* Last()
* LastOrDefault()
* Single()
* SingleOrDefault()

**Comment by briankavanaugh** [comment]
Related to this is a false positive. Because it's not flagging item as possibly null, if you check for item != null, it will flag that as always evaluating to true.

```
string DoSomething(IEnumerable<object> list )
{
    var item = list.FirstOrDefault();
    if(item != null) // flagged by https://rules.sonarsource.com/csharp/RSPEC-2589
        return item.ToString();
    return null;
}

**Comment by pavel-mikula-sonarsource** [comment]
Hi @briankavanaugh ,

Yes, you're right. Same underlying problem will effect all rules based on Symbolic Execution engine.

**Comment by andrewgierens** [comment]
Hello! Is there a work around for this? Would be really beneficial to have this as we utilise linq very heavily.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @andrewgierens, 

There's unfortunately no workaround. We'll rework our Symbolic Execution engine in near future and that will allow us to fix this.
